containers/datalab/run-with-gce.sh

```diff
@@ -60,14 +61,39 @@ fi
 
 echo "${DOCS}"
 
-INSTANCE=`gcloud compute instances list --regex "datalab-kernel-gateway-[0-9]*" --limit 1 --format "value(name)"`
+# We want to run the kernel gateway in a VM. However, we also want to make sure
+# there is a 1:1 correspondence between end user and kernel gateway, so we need
+# to create a separate VM for each user.
+#
+# To do this, we will name the VM with a prefix tied to the user. Since
+# usernames can include arbitrary characters, we sanitize the username by first
+# hashing it. We use the openssl implementation of sha1 for this purpose,
+# solely for the fact that it should be nearly ubiquitous.
+#
+# NOTE: THIS IS NOT A SECURITY SANDBOX. When running in a GCE VM, Datalab is in
+# an inherently shared environment. This separation is merely to prevent
+# accidental conflicts between users, not to provide any sort of privacy or
+# authentication.
+USER_EMAIL=`gcloud info --format="value(config.account)"`
+if [[ -z "${USER_EMAIL}" ]]; then
+  echo "You must log in via the gcloud tool"
+  echo "    gcloud auth login"
+  exit ${ERR_LOGIN}
+fi
+
+echo "Looking up gateway VM for ${USER_EMAIL}..."
+
+USER_HASH=`echo "${USER_EMAIL}" | openssl sha1 -r | cut -d ' ' -f 1`
```

Comment:
It seems openssl is not built-in in Windows? If so, can we use pure string replacement to sanitize the email to remove the dependency? Or maybe you plan to create a different script for end users (not for Datalab developers)?